In [ ]:
#importing libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import matplotlib.patches as mpatches
import seaborn as sns
import warnings 
warnings.filterwarnings("ignore")

In [ ]:
!gdown --id 1aQC8kOD1bL5As_LK30NfRQcxBNVKMZqF

Downloading...
From: https://drive.google.com/uc?id=1aQC8kOD1bL5As_LK30NfRQcxBNVKMZqF
To: /content/vehicles.pickle
922MB [00:13, 67.0MB/s]


In [ ]:
df=pd.read_csv('vehicles.csv')#reading the dataframe using pandas inbuilt read_csv function

In [ ]:
df.head()

In [ ]:
#dropping off irrelevant features like Unnamed:0,id,url,region_url,image_url,posting_date,VIN
vehicles=df.drop(['Unnamed: 0','id','url','region_url','image_url','posting_date','VIN'],axis=1)

In [ ]:
import warnings
warnings.filterwarnings("ignore")
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from subprocess import check_output
%matplotlib inline
import plotly.offline as py
py.init_notebook_mode(connected=True)
import plotly.graph_objs as go
import plotly.tools as tls
import os
import gc

import nltk
import re
from nltk.corpus import stopwords
from nltk.stem import PorterStemmer
from bs4 import BeautifulSoup
import re
from nltk.corpus import stopwords
# This package is used for finding longest common subsequence between two strings
# you can write your own dp code for this

from nltk.stem import PorterStemmer
from bs4 import BeautifulSoup
from sklearn.manifold import TSNE
# Import the Required lib packages for WORD-Cloud generation
# https://stackoverflow.com/questions/45625434/how-to-install-wordcloud-in-python3-6
from wordcloud import WordCloud, STOPWORDS
from os import path
from PIL import Image


In [ ]:
nltk.download('stopwords')


[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


True

Data Preprocessing for description feature

In [ ]:
STOP_WORDS = stopwords.words("english")


def preprocess(x):
    x = str(x).lower()
    x = x.replace(",000,000", "m").replace(",000", "k").replace("′", "'").replace("’", "'")\
                           .replace("won't", "will not").replace("cannot", "can not").replace("can't", "can not")\
                           .replace("n't", " not").replace("what's", "what is").replace("it's", "it is")\
                           .replace("'ve", " have").replace("i'm", "i am").replace("'re", " are")\
                           .replace("he's", "he is").replace("she's", "she is").replace("'s", " own")\
                           .replace("%", " percent ").replace("₹", " rupee ").replace("$", " dollar ")\
                           .replace("€", " euro ").replace("'ll", " will")
    x = re.sub(r"([0-9]+)000000", r"\1m", x)
    x = re.sub(r"([0-9]+)000", r"\1k", x)
    
    
    porter = PorterStemmer()
    pattern = re.compile('\W')
    
    if type(x) == type(''):
        x = re.sub(pattern, ' ', x)
    
    
    if type(x) == type(''):
        x = porter.stem(x)
        example1 = BeautifulSoup(x)
        x = example1.get_text()
               
    
    return x
    

In [ ]:
vehicles['description']=vehicles['description'].apply(preprocess)

In [ ]:
vehicles['description']=vehicles['description'].apply(lambda words: ' '.join(word.lower() for word in words.split() if word not in STOP_WORDS))

In [ ]:
vehicles['description']=vehicles['description'].apply(lambda x:re.sub('[^A-Za-z0-9]+', ' ',x))

In [ ]:
vehicles['description']=vehicles['description'].apply(lambda x:re.sub(r'[0-9]+', '',x))

In [ ]:
vehicles['description']=vehicles['description'].apply(lambda x:re.sub(' +', ' ', x))


In [ ]:
vehicles.loc[2,'description']

'clean bmw x i beautiful rare blue water metallic exterior tan interior color combination speed automatic transmission awd cd fm radio cold c serviced along w oil change alloy wheels split rear seats driver passenger air bags multi function remote w keyless entry electric windows door locks cruise control lighted vanity mirrors many extras missing tow eye cover rear dollar replace tires percent tread remaining blemishes exterior scuffs scratches normal wear tear would make excellent safe run around town college vehicle title hand priced sell'

In [ ]:
import pickle
with open('vehicles.pickle', 'wb') as handle:
    pickle.dump(vehicles, handle)

In [ ]:
import phik
from phik import resources, report

In [ ]:
import pickle
inifile=open('vehicles.pickle','rb')
vehicles=pickle.load(inifile)


Correlation matrix b/w the features

In [ ]:
vehicles.drop('description',axis=1).phik_matrix()

interval columns not set, guessing: ['price', 'year', 'odometer', 'lat', 'long']


/usr/local/lib/python3.7/dist-packages/phik/data_quality.py:60: UserWarning: The number of unique values of variable model is large: 31520. Are you sure this is not an interval variable? Analysis for pairs of variables including model can be slow.
  .format(col, df[col].nunique())


region     price      year  ...     state       lat      long
region        1.000000  0.007235  0.168316  ...  0.999951  0.839512  0.864222
price         0.007235  1.000000  0.025356  ...  0.000000  0.000000  0.000000
year          0.168316  0.025356  1.000000  ...  0.092438  0.032187  0.032970
manufacturer  0.359853  0.000000  0.260431  ...  0.237155  0.065319  0.101932
model         0.930909  0.000000  0.975571  ...  0.828284  0.722142  0.792295
condition     0.338174  0.000000  0.210524  ...  0.192519  0.056490  0.097790
cylinders     0.252824  0.000000  0.157098  ...  0.142315  0.039329  0.056118
fuel          0.257021  0.001505  0.129585  ...  0.152078  0.038682  0.053019
odometer      0.000000  0.707105  0.000000  ...  0.000000  0.000000  0.000000
title_status  0.248687  0.000000  0.155263  ...  0.160017  0.028926  0.041467
transmission  0.289072  0.000000  0.334645  ...  0.191406  0.015087  0.129334
drive         0.420982  0.000000  0.278891  ...  0.375649  0.153584  0.101113
size          0.281284  0.000000  0.074196  ...  0.183806  0.044856  0.078748
type          0.360340  0.000000  0.207356  ...  0.232015  0.055320  0.091765
paint_color   0.203385  0.000000  0.182834  ...  0.129731  0.028817  0.053136
state         0.999951  0.000000  0.092438  ...  1.000000  0.814421  0.842970
lat           0.839512  0.000000  0.032187  ...  0.814421  1.000000  0.615750
long          0.864222  0.000000  0.032970  ...  0.842970  0.615750  1.000000

[18 rows x 18 columns]

In [ ]:
#price is highly correlated to odometer so odometer is the most imp feature
#model has almost very high correlation with every other feature.
#state is highly correlated to region,lat,long.



In [ ]:
dropping highly correlated features. i.e State,Region,model

In [ ]:
vehicles=vehicles.drop(['state','region','model'],axis=1)

Randomly filling null values

In [ ]:
import random
for i in vehicles.columns:
  vehicles[i]=vehicles[i].fillna(random.choice(vehicles[vehicles[i] != np.nan][i]))

Train-Test Split

In [ ]:
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split

In [ ]:
X=vehicles.drop('price',axis=1)
y=vehicles['price']

In [ ]:
x_train,x_test,y_train,y_test=train_test_split(X,y,test_size =0.2,random_state=21)

Scaling the numerical Columns


In [ ]:
scaler=StandardScaler()

In [ ]:
numerical_columns=['year','odometer','lat','long']
categorical_columns=[]
for i in X.columns:
    if i not in numerical_columns:
        categorical_columns.append(i)

In [ ]:
X['year']=X['year'].astype('int64')

In [ ]:
#categorical_columns.remove('description')
categorical_columns

['manufacturer',
 'condition',
 'cylinders',
 'fuel',
 'title_status',
 'transmission',
 'drive',
 'size',
 'type',
 'paint_color',
 'description']

In [ ]:
for i in numerical_columns:
  x_train[i]=scaler.fit_transform(x_train[i].values.reshape(-1,1))

In [ ]:
for i in numerical_columns:
  x_test[i]=scaler.transform(x_test[i].values.reshape(-1,1))

In [ ]:
x_train.shape,x_test.shape

((366570, 15), (91643, 15))

Embedding description text using TFIDF Vectorizer

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.feature_extraction.text import CountVectorizer

In [ ]:
vectorizer = TfidfVectorizer(min_df=5,ngram_range=(1,3), max_features=5000)#max_features=5000 mean dictionary will consist max 10000 words

In [ ]:
vectorizer.fit(x_train['description'].values)#fitting X-train['description'] to create dictionary of unique words


TfidfVectorizer(analyzer='word', binary=False, decode_error='strict',
                dtype=<class 'numpy.float64'>, encoding='utf-8',
                input='content', lowercase=True, max_df=1.0, max_features=5000,
                min_df=5, ngram_range=(1, 3), norm='l2', preprocessor=None,
                smooth_idf=True, stop_words=None, strip_accents=None,
                sublinear_tf=False, token_pattern='(?u)\\b\\w\\w+\\b',
                tokenizer=None, use_idf=True, vocabulary=None)

In [ ]:
x_train_description = vectorizer.transform(x_train['description'].values)#applying tranform function to X_train,X_test


In [ ]:
x_test_description = vectorizer.transform(x_test['description'].values)

In [ ]:
x_train_description.shape,x_test_description.shape

((366570, 5000), (91643, 5000))

One hot encoding on remaining Categorical Columns

In [ ]:
vectorizer=CountVectorizer(binary=True)
x_train_condition_ohe=vectorizer.fit_transform(x_train['condition'].astype('str').values)
x_test_condition_ohe=vectorizer.transform(x_test['condition'].astype('str').values)

In [ ]:
vectorizer=CountVectorizer(binary=True)
x_train_cylinders_ohe=vectorizer.fit_transform(x_train['cylinders'].astype('str').values)
x_test_cylinders_ohe=vectorizer.transform(x_test['cylinders'].astype('str').values)




In [ ]:
vectorizer=CountVectorizer(binary=True)
x_train_fuel_ohe=vectorizer.fit_transform(x_train['fuel'].values)
x_test_fuel_ohe=vectorizer.transform(x_test['fuel'].values)




In [ ]:
vectorizer=CountVectorizer(binary=True)
x_train_title_ohe=vectorizer.fit_transform(x_train['title_status'].values)
x_test_title_ohe=vectorizer.transform(x_test['title_status'].values)




In [ ]:
vectorizer=CountVectorizer(binary=True)
x_train_transmission_ohe=vectorizer.fit_transform(x_train['transmission'].values)
x_test_transmission_ohe=vectorizer.transform(x_test['transmission'].values)




In [ ]:

vectorizer=CountVectorizer(binary=True)
x_train_type_ohe=vectorizer.fit_transform(x_train['type'].astype('str').values)
x_test_type_ohe=vectorizer.transform(x_test['type'].astype('str').values)




In [ ]:
vectorizer=CountVectorizer(binary=True)
x_train_drive_ohe=vectorizer.fit_transform(x_train['drive'].astype('str').values)
x_test_drive_ohe=vectorizer.transform(x_test['drive'].astype('str').values)




In [ ]:
vectorizer=CountVectorizer(binary=True)
x_train_paint_color_ohe=vectorizer.fit_transform(x_train['paint_color'].astype('str').values)
x_test_paint_color_ohe=vectorizer.transform(x_test['paint_color'].astype('str').values)




In [ ]:
vectorizer=CountVectorizer(binary=True)
x_train_size_ohe=vectorizer.fit_transform(x_train['size'].astype('str').values)
x_test_size_ohe=vectorizer.transform(x_test['size'].astype('str').values)




In [ ]:
vectorizer=CountVectorizer(binary=True)
x_train_mf_ohe=vectorizer.fit_transform(x_train['manufacturer'].astype('str').values)
x_test_mf_ohe=vectorizer.transform(x_test['manufacturer'].astype('str').values)




In [ ]:
x_odometer_normalize=x_train['odometer'].values.reshape(-1,1)
x_year_normalize=x_train['year'].values.reshape(-1,1)
x_lat_normalize=x_train['lat'].values.reshape(-1,1)
x_long_normalize=x_train['long'].values.reshape(-1,1)


In [ ]:
x_odometer_normalize_test=x_test['odometer'].values.reshape(-1,1)
x_year_normalize_test=x_test['year'].values.reshape(-1,1)
x_lat_normalize_test=x_test['lat'].values.reshape(-1,1)
x_long_normalize_test=x_test['long'].values.reshape(-1,1)


In [ ]:
from scipy.sparse import hstack
X_train_total=hstack((x_train_condition_ohe,x_train_cylinders_ohe,x_train_fuel_ohe,x_train_title_ohe,x_train_transmission_ohe,x_train_type_ohe,x_train_drive_ohe,x_train_paint_color_ohe,x_train_mf_ohe,x_train_size_ohe,x_train_description,x_odometer_normalize,x_year_normalize,x_lat_normalize,x_long_normalize)).tocsr()

In [ ]:
X_test_total=hstack((x_test_condition_ohe,x_test_cylinders_ohe,x_test_fuel_ohe,x_test_title_ohe,x_test_transmission_ohe,x_test_type_ohe,x_test_drive_ohe,x_test_paint_color_ohe,x_test_mf_ohe,x_test_size_ohe,x_test_description,x_odometer_normalize_test,x_year_normalize_test,x_lat_normalize_test,x_long_normalize_test)).tocsr()

In [ ]:
with open('x_train_final.pickle', 'wb') as handle:
    pickle.dump(X_train_total, handle)
with open('x_test_final.pickle', 'wb') as handle:
    pickle.dump(X_test_total, handle)

Models:

In [ ]:
from sklearn.linear_model import LinearRegression
from sklearn.linear_model import SGDRegressor
from sklearn.svm import SVR
from sklearn.linear_model import Lasso
from sklearn.linear_model import Ridge

In [ ]:
X_train_total.shape,X_test_total.shape

((366570, 5111), (91643, 5111))

In [ ]:
type(X_train_total)

scipy.sparse.csr.csr_matrix

In [ ]:
import pickle
inifile=open('x_train_final.pickle','rb')
X_train_total=pickle.load(inifile)
inifile=open('x_test_final.pickle','rb')
X_test_total=pickle.load(inifile)


The size and dimensionality of dataset is very large so we will reduce the dimensionality using PCA .

In [ ]:
from sklearn.decomposition import TruncatedSVD
pca = TruncatedSVD(n_components = 200)
pca.fit(X_train_total)
X_train_reduced = pca.transform(X_train_total)

In [ ]:
X_test_reduced = pca.transform(X_test_total)

In [ ]:
pca.explained_variance_ratio_.sum()#total variance explained after reduction

0.9477639915612277

In [ ]:
X_train_reduced.shape

(366570, 50)

Linear Regression

In [ ]:
LR=LinearRegression()
LR.fit(X_train_reduced,y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [ ]:
from sklearn.metrics import mean_absolute_error
from sklearn.metrics import r2_score

In [ ]:
y_pred=LR.predict(X_test_reduced)
mean_absolute_error(y_test,y_pred)

2540467.7540413956

In [ ]:
r2_score(y_test,y_pred)

-1.5793750283665826

SGD Regression


In [ ]:
sgd=SGDRegressor()
sgd.fit(X_train_reduced,y_train)

SGDRegressor(alpha=0.0001, average=False, early_stopping=False, epsilon=0.1,
             eta0=0.01, fit_intercept=True, l1_ratio=0.15,
             learning_rate='invscaling', loss='squared_loss', max_iter=1000,
             n_iter_no_change=5, penalty='l2', power_t=0.25, random_state=None,
             shuffle=True, tol=0.001, validation_fraction=0.1, verbose=0,
             warm_start=False)

In [ ]:
y_pred=sgd.predict(X_test_reduced)
mean_absolute_error(y_test,y_pred)

646423653456494.8

In [ ]:
r2_score(y_test,y_pred)

-6.340478588099479e+16

Lasso Regression

In [ ]:
ls=Lasso()
ls.fit(X_train_reduced,y_train)

Lasso(alpha=1.0, copy_X=True, fit_intercept=True, max_iter=1000,
      normalize=False, positive=False, precompute=False, random_state=None,
      selection='cyclic', tol=0.0001, warm_start=False)

In [ ]:
y_test_pred=ls.predict(X_test_reduced)
mean_absolute_error(y_test,y_test_pred)

2537123.805504272

In [ ]:
r2_score(y_test,y_test_pred)

-1.575091886881328

Ridge Regression

In [ ]:
rd=Ridge()
rd.fit(X_train_reduced,y_train)

Ridge(alpha=1.0, copy_X=True, fit_intercept=True, max_iter=None,
      normalize=False, random_state=None, solver='auto', tol=0.001)

In [ ]:
y_test_pred=rd.predict(X_test_reduced)
mean_absolute_error(y_test,y_test_pred)


2542873.8608324835

In [ ]:
r2_score(y_test,y_test_pred)


-1.5810075714599185

So linear models are giving r2_score as 1.57 or 1.58 . This is the worst score we can have . 

In [ ]:
 !gdown --id 1qKND6Iwqgb-wU_oAPP1W9B_K-noTvN6d

Downloading...
From: https://drive.google.com/uc?id=1qKND6Iwqgb-wU_oAPP1W9B_K-noTvN6d
To: /content/x_test_final.pickle
307MB [00:06, 47.2MB/s]


In [ ]:
 !gdown --id 1HHQueME8w7fSYBOmA4t8-EURHQF6ddGy

Downloading...
From: https://drive.google.com/uc?id=1HHQueME8w7fSYBOmA4t8-EURHQF6ddGy
To: /content/x_train_final.pickle
1.23GB [00:15, 80.8MB/s]


In [ ]:
import pickle
inifile=open('x_train_final.pickle','rb')
X_train_total=pickle.load(inifile)


In [ ]:
inifile=open('x_test_final.pickle','rb')
X_test_total=pickle.load(inifile)


In [ ]:
from sklearn.tree import DecisionTreeRegressor

In [ ]:
dt=DecisionTreeRegressor()

In [ ]:
dt.fit(X_train_reduced,y_train)

DecisionTreeRegressor(ccp_alpha=0.0, criterion='mse', max_depth=None,
                      max_features=None, max_leaf_nodes=None,
                      min_impurity_decrease=0.0, min_impurity_split=None,
                      min_samples_leaf=1, min_samples_split=2,
                      min_weight_fraction_leaf=0.0, presort='deprecated',
                      random_state=None, splitter='best')

In [ ]:
y_test_pred=dt.predict(X_test_reduced)
mean_absolute_error(y_test,y_test_pred)


108055447.78468622

In [ ]:
r2_score(y_test,y_test_pred)


-14905.49671727345

In [ ]:
import xgboost
from xgboost import XGBRegressor


In [ ]:
xgb=XGBRegressor(max_depth=3,n_estimators=100)

In [ ]:
xgb.fit(X_train_reduced,y_train)

[10:28:19] WARNING: /workspace/src/objective/regression_obj.cu:152: reg:linear is now deprecated in favor of reg:squarederror.


XGBRegressor(base_score=0.5, booster='gbtree', colsample_bylevel=1,
             colsample_bynode=1, colsample_bytree=1, gamma=0,
             importance_type='gain', learning_rate=0.1, max_delta_step=0,
             max_depth=3, min_child_weight=1, missing=None, n_estimators=100,
             n_jobs=1, nthread=None, objective='reg:linear', random_state=0,
             reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=None,
             silent=None, subsample=1, verbosity=1)

In [ ]:
y_test_pred=xgb.predict(X_test_reduced)
mean_absolute_error(y_test,y_test_pred)


14638148.002295934

In [ ]:
r2_score(y_test,y_test_pred)


-51.163696152078124